applications/nrf5340_audio/src/audio/sw_codec_select.c

```diff
@@ -206,109 +235,89 @@ int sw_codec_decode(uint8_t const *const encoded_data, size_t encoded_size, bool
+			audio_data_tx.meta.data_len_us = CONFIG_AUDIO_FRAME_DURATION_US;
+			audio_data_tx.meta.sample_rate_hz = m_config.decoder.sample_rate_hz;
+			audio_data_tx.meta.bits_per_sample = CONFIG_AUDIO_BIT_DEPTH_BITS;
+			audio_data_tx.meta.carried_bits_pr_sample = CONFIG_AUDIO_BIT_DEPTH_BITS;
+			if (m_config.decoder.channel_mode == SW_CODEC_MONO) {
+				audio_data_tx.meta.locations = 0x00000001;
+			} else {
+				audio_data_tx.meta.locations = 0x00000003;
+			}
+			audio_data_tx.meta.bad_data = bad_frame;
+
+			audio_data_rx.data = (void *)decoded_data_stereo;
+			audio_data_rx.data_size = PCM_NUM_BYTES_STEREO;
+			audio_data_rx.meta.data_coding = PCM;
+			audio_data_rx.meta.data_len_us = CONFIG_AUDIO_FRAME_DURATION_US;
+			audio_data_rx.meta.sample_rate_hz = m_config.decoder.sample_rate_hz;
+			audio_data_rx.meta.bits_per_sample = CONFIG_AUDIO_BIT_DEPTH_BITS;
+			audio_data_rx.meta.carried_bits_pr_sample = CONFIG_AUDIO_BIT_DEPTH_BITS;
+			audio_data_rx.meta.locations = audio_data_tx.meta.locations;
+
+			ret = audio_module_data_tx_rx(&handle[LC3_MODULE_ID_DECODER],
+						      &handle[LC3_MODULE_ID_DECODER],
+						      &audio_data_tx, &audio_data_rx, K_FOREVER);
 			if (ret) {
 				return ret;
 			}
-			break;
 		}
-		case SW_CODEC_STEREO: {
-			if (bad_frame && IS_ENABLED(CONFIG_SW_CODEC_OVERRIDE_PLC)) {
-				memset(decoded_data_mono[AUDIO_CH_L], 0, PCM_NUM_BYTES_MONO);
-				memset(decoded_data_mono[AUDIO_CH_R], 0, PCM_NUM_BYTES_MONO);
-				decoded_data_size = PCM_NUM_BYTES_MONO;
-			} else {
-				/* Decode left channel */
-				ret = sw_codec_lc3_dec_run(
-					encoded_data, encoded_size / 2, LC3_PCM_NUM_BYTES_MONO,
-					AUDIO_CH_L, decoded_data_mono[AUDIO_CH_L],
-					(uint16_t *)&decoded_data_size, bad_frame);
-				if (ret) {
-					return ret;
-				}
-
-				/* Decode right channel */
-				ret = sw_codec_lc3_dec_run(
-					(encoded_data + (encoded_size / 2)), encoded_size / 2,
-					LC3_PCM_NUM_BYTES_MONO, AUDIO_CH_R,
-					decoded_data_mono[AUDIO_CH_R],
-					(uint16_t *)&decoded_data_size, bad_frame);
-				if (ret) {
-					return ret;
-				}
-
-				for (int i = 0; i < m_config.decoder.channel_mode; ++i) {
-					ret = sw_codec_sample_rate_convert(
-						&decoder_converters[i],
-						m_config.decoder.sample_rate_hz,
-						CONFIG_AUDIO_SAMPLE_RATE_HZ, decoded_data_mono[i],
-						decoded_data_size,
-						decoded_data_mono_system_sample_rate[i],
-						&pcm_in_data_ptrs[i], &pcm_size_mono);
-					if (ret) {
-						LOG_ERR("Sample rate conversion failed for channel "
-							"%d : %d",
-							i, ret);
-						return ret;
-					}
-				}
-			}
 
-			ret = pscm_combine(pcm_in_data_ptrs[AUDIO_CH_L],
-					   pcm_in_data_ptrs[AUDIO_CH_R], pcm_size_mono,
-					   CONFIG_AUDIO_BIT_DEPTH_BITS, pcm_data_stereo,
-					   &pcm_size_stereo);
+		dec_data_stereo = decoded_data_stereo;
+		dec_data_stereo_src = decoded_data_stereo_system_sample_rate;
+
+		for (int i = 0; i < m_config.decoder.num_ch; ++i) {
+			ret = sw_codec_sample_rate_convert(
```

Comment:
Would it be within the `audio_module` to either re-write the converter or write a wrapper that allows the sample rate converter to work on blocks of `struct audio_data`?

Reply:
I would probably add this as a module and then use it as a sub-module to the decoder and encoder, so yes add a wrapper for audio_data. This would then give a dec/enc module that returns audio data at the output sample rate. Alternatively the resampler could be an audio module in it's own right, but I am unclear as to the overhead of having multiple threads.